File: c++/20529.cpp
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

vector<string> mbtis;
vector<string> selected;
int minDist;

void init() {
  minDist = 20;
  while (!mbtis.empty()) {
    mbtis.pop_back();
  }
}

void dfs(int idx) {
  if (selected.size() == 3) {
    int dist = 0;
    for (int i = 0; i < 4; i++) {
      if (selected[0][i] != selected[1][i]) dist++;
      if (selected[0][i] != selected[2][i]) dist++;
      if (selected[1][i] != selected[2][i]) dist++;
    }
    minDist = min(minDist, dist);
    return;
  }
  for (int i = idx; i < mbtis.size(); i++) {
    selected.push_back(mbtis[i]);
    dfs(i + 1);
    selected.pop_back();
  }
}

int main() {
  freopen("example.txt", "r", stdin);
  int T;
  cin >> T;

  for (int i = 0; i < T; i++) {
    int N;
    cin >> N;

    init();

    for (int j = 0; j < N; j++) {
      string s;
      cin >> s;
      mbtis.push_back(s);
    }

    if (N > 32) {
      cout << 0 << '\n';
      continue;
    }

    dfs(0);

    cout << minDist << '\n';
  }

  return 0;
}

File: js/20529.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

const T = +inputs[0];

const answer = [];

for (let i = 1; i < T * 2; i += 2) {
  const N = inputs[i];
  const mbtis = inputs[i + 1].split(' ');

  if (N > 32) {
    answer.push(0);
    continue;
  }

  const selected = [];
  let minDist = 20;

  dfs(0, 0);
  answer.push(minDist);

  function dfs(idx, cnt) {
    if (cnt === 3) {
      let dist = 0;
      for (let i = 0; i < 4; i++) {
        if (selected[0][i] !== selected[1][i]) dist++;
        if (selected[0][i] !== selected[2][i]) dist++;
        if (selected[1][i] !== selected[2][i]) dist++;
      }
      minDist = Math.min(minDist, dist);
      return;
    }
    for (let i = idx; i < mbtis.length; i++) {
      selected.push(mbtis[i]);
      dfs(i + 1, cnt + 1);
      selected.pop();
    }
  }
}

console.log(answer.join('\n'));
